python/cairo-addons/src/cairo_addons/testing/hints.py

```diff
@@ -28,111 +28,109 @@ def _debug_info(pc):
-            ...
+        programs: A list of Cairo programs (either `Program` or `RustProgram`).
+        hint: The original hint code to replace.
+        new_hint: The new hint code to use.
+        scope: Optional scope name to restrict which hints are patched.
 
-        # Replace a regular hint
-        with patch_hint(program, 'ids.x = 5', 'ids.x = 10'):
-            ...
+    Yields:
+        The programs with the hint patched (same list for `Program`, new list for `RustProgram`).
 
-        # Replace hint only in specific scope
-        with patch_hint(program, 'ids.x = 5', 'ids.x = 10', scope='my_function'):
-            ...
+    Raises:
+        ValueError: If the hint is not found in any program.
+        TypeError: If the program type is unsupported.
     """
+    import contextlib
     import re
 
-    def get_nondet_arg(hint_code: str) -> Optional[str]:
-        """Extract argument from nondet hint if it is one, otherwise return None."""
-        if match := re.match(r"nondet %{(.+)%};", hint_code.strip()):
-            return match.group(1).strip()
-        return None
-
-    def parse_fp_assignment_hint(hint_code: str) -> tuple[Optional[str], Optional[str]]:
-        """Extract memory location and argument from an fp assignment hint."""
-        if match := re.match(
-            r"memory\[(.+)\] = to_felt_or_relocatable\((.+)\)", hint_code
-        ):
-            return match.group(1), match.group(2).strip()
-        return None, None
-
-    orig_nondet_arg = get_nondet_arg(hint)
-    new_nondet_arg = get_nondet_arg(new_hint)
-
-    patched_programs_hints = [{} for _ in programs]
-
-    for i, program in enumerate(programs):
-        patched_hints = {}
-        for k, hint_list in program.hints.items():
-            new_hints = []
-            for hint_ in hint_list:
-                # Skip hints not in specified scope
-                if scope is not None and scope not in str(hint_.accessible_scopes[-1]):
-                    new_hints.append(hint_)
-                    continue
-
-                if orig_nondet_arg:
-                    # Handle nondet hint patching
-                    mem_loc, arg = parse_fp_assignment_hint(hint_.code)
-                    if arg == orig_nondet_arg:
-                        new_hints.append(
-                            CairoHint(
-                                accessible_scopes=hint_.accessible_scopes,
-                                flow_tracking_data=hint_.flow_tracking_data,
-                                code=f"memory[{mem_loc}] = to_felt_or_relocatable({new_nondet_arg})",
-                            )
-                        )
-                    else:
-                        new_hints.append(hint_)
-                else:
-                    # Handle regular hint patching
-                    if hint_.code.strip() == implementations.get(
-                        hint.strip(), hint.strip()
+    if not programs:
+        raise ValueError("No programs provided")
```

Comment:
I don't think we should validate that, the arg type does not allow for None anyway